### PR TITLE
fix(sqlite): preserve GREATEST/LEAST NULL semantics

### LINKED ISSUE #8315: GREATEST/LEAST → SQLite MAX/MIN silently changes NULL semantics
`GREATEST`/`LEAST` ignore NULL arguments in Postgres and DuckDB. SQLite's multi-argument
`max()`/`min()` return NULL if *any* argument is NULL. sqlglot renames one to the other with
no compensation, so the translated query runs successfully and returns different values.

This is the same bug class as #8283 (`CONCAT` → `||`), which was fixed by routing through
`convert_concat_args`. `GREATEST`/`LEAST` have no equivalent path.

**Read dialect:** `postgres` (also reproduces from `duckdb`)
**Write dialect:** `sqlite`
**Still on main:** yes — `sqlglot/generators/sqlite.py` `least_sql`/`greatest_sql` are a bare
`rename_func("MIN")`/`rename_func("MAX")`. Verified on 30.18.0.

### Fully reproducible code snippet

```python
import sqlite3, sqlglot

sql = "SELECT GREATEST(a, b) AS hi, LEAST(a, b) AS lo FROM t"
out = sqlglot.transpile(sql, read="postgres", write="sqlite")[0]
print(out)          # SELECT MAX(a, b) AS hi, MIN(a, b) AS lo FROM t

con = sqlite3.connect(":memory:")
con.execute("CREATE TABLE t (a INTEGER, b INTEGER)")
con.execute("INSERT INTO t VALUES (NULL, 1)")
print(con.execute(out).fetchall())   # [(None, None)]
                                     # Postgres/DuckDB give [(1, 1)]
```

| | `GREATEST(a, b)` | `LEAST(a, b)` |
|---|---|---|
| Postgres / DuckDB | `1` | `1` |
| sqlglot's SQLite output | `NULL` | `NULL` |

### Official documentation

- Postgres, [GREATEST and LEAST](https://www.postgresql.org/docs/current/functions-conditional.html#FUNCTIONS-GREATEST-LEAST):
  > NULL values in the argument list are ignored. The result will be NULL only if all the
  > expressions evaluate to NULL. (This is a deviation from the SQL standard. According to the
  > standard, the return value is NULL if any argument is NULL. Some other databases behave this way.)
- SQLite, [max()](https://www.sqlite.org/lang_corefunc.html#max_scalar):
  > The multi-argument max() function returns the argument with the maximum value, or return
  > NULL if any argument is NULL.

  `min()` behaves symmetrically.

### Possible emission

`max()`/`min()` in *aggregate* position skip NULLs, but that form can't be used here — a
subquery in `FROM` can't reference the outer row's columns. Replacing each argument with a
rotation of `COALESCE` over the argument list does work, and needs no new SQLite features:

```sql
-- LEAST(a, b)
MIN(COALESCE(a, b), COALESCE(b, a))
-- GREATEST(a, b, c)
MAX(COALESCE(a, b, c), COALESCE(b, c, a), COALESCE(c, a, b))
```

### PR DESCRIPTION
**Pull request by Kunal8954**
Fixes https://github.com/tobymao/sqlglot/issues/8315

## Summary

Fix SQLite transpilation of `GREATEST` and `LEAST` when the source dialect ignores NULL arguments.

SQLite's multi-argument `MAX` and `MIN` return NULL when any argument is NULL, while PostgreSQL and DuckDB `GREATEST`/`LEAST` ignore NULL arguments and return NULL only when all arguments are NULL.

The SQLite generator now honors SQLGlot's existing `ignore_nulls` semantic flag. For dialects where `GREATEST`/`LEAST` ignore NULL values, it generates a COALESCE rotation before applying SQLite `MAX`/`MIN`.

For example:

```sql
-- SQLite does not have GREATEST/LEAST; previously GREatEST was renamed to MAX,
-- silently changing NULL semantics.

-- postgres: SELECT GREATEST(a, b)
SELECT MAX(COALESCE(a, b), COALESCE(b, a))

-- postgres: SELECT LEAST(a, b, c)
SELECT MIN(COALESCE(a, b, c), COALESCE(b, c, a), COALESCE(c, a, b))
```

Each argument is the first argument of exactly one COALESCE in the rotation, so every non-NULL argument is reproduced and compared, while an all-NULL input still yields NULL. This matches PostgreSQL/DuckDB semantics exactly (verified by executing the emitted SQLite SQL).

When the source dialect already returns NULL if any argument is NULL (e.g. MySQL), the previous `MAX`/`MIN` rename is kept unchanged.

## Tests

- Updated the two pre-existing `postgres -> sqlite` assertions in `tests/dialects/test_sqlite.py` that encoded the buggy output.
- Added coverage for Postgres and DuckDB readers, two- and three-argument calls, and literal NULL values.
- All new assertions fail against the previous implementation and pass with this fix.
- Verified at runtime with `sqlite3`: `GREATEST(NULL, 1)` and `LEAST(NULL, 1)` now evaluate to `1` (previously `NULL`).

### CODE REVIEW

File: sqlglot/generators/sqlite.py [hunk 1]
```diff
@@ -338,17 +338,36 @@ def groupconcat_sql(self, expression: exp.GroupConcat) -> str:
...
-        if expression.expressions:
-            return rename_func("MIN")(self, expression)
-
-        return self.sql(expression, "this")
+    def _greatest_least_sql(self, expression: exp.Greatest | exp.Least) -> str:
+        if not expression.expressions:
+            return self.sql(expression, "this")
+
+        if not expression.args.get("ignore_nulls"):
+            name = "MAX" if isinstance(expression, exp.Greatest) else "MIN"
```
Comment (treysp) [inline sqlglot/generators/sqlite.py]: Let's pull this out of the if block and use it for both returns

File: sqlglot/generators/sqlite.py [hunk 2]
```diff
@@ -338,17 +338,36 @@ def groupconcat_sql(self, expression: exp.GroupConcat) -> str:
...
+
+        # GREATEST/LEAST ignore NULLs, but SQLite's multi-argument MAX/MIN
+        # return NULL if any argument is NULL. Emit a rotation of COALESCE
+        # calls so that every argument is the first argument of exactly one
+        # COALESCE, which then returns it if it's non-NULL. Comparing all of
+        # those yields the ignored-NULLs result, and an all-NULL input still
+        # yields NULL:
+        # GREATEST(a, b, c) -> MAX(COALESCE(a, b, c), COALESCE(b, c, a), COALESCE(c, a, b)).
+        args = [expression.this, *expression.expressions]
+        aggregate = exp.Max if isinstance(expression, exp.Greatest) else exp.Min
```
Comment (treysp) [inline sqlglot/generators/sqlite.py]: delete - not needed when using self.func below

File: sqlglot/generators/sqlite.py [hunk 3]
```diff
@@ -338,17 +338,36 @@ def groupconcat_sql(self, expression: exp.GroupConcat) -> str:
...
+
+        coalesces = []
+        for i in range(len(args)):
+            rotated = args[i:] + args[:i]
+            coalesces.append(exp.Coalesce(this=rotated[0], expressions=rotated[1:]))
+
+        return self.sql(aggregate(expressions=coalesces))
```
Comment (treysp) [inline sqlglot/generators/sqlite.py]:
```suggestion
        return self.func(name, *coalesces)
```

File: tests/dialects/test_sqlite.py [hunk 4]
```diff
@@ -127,17 +127,44 @@ def test_sqlite(self):
...
+        # Literal NULLs exercise the ignore-NULLs semantics: a NULL anywhere in
+        # the arguments must be skipped and only an all-NULL input returns NULL.
+        self.validate_all(
+            "SELECT MAX(COALESCE(NULL, 1), COALESCE(1, NULL))",
+            read={"postgres": "SELECT GREATEST(NULL, 1)"},
+        )
+        self.validate_all(
+            "SELECT MIN(COALESCE(3, NULL, 1), COALESCE(NULL, 1, 3), COALESCE(1, 3, NULL))",
+            read={"postgres": "SELECT LEAST(3, NULL, 1)"},
+        )
```
Comment (treysp) [inline tests/dialects/test_sqlite.py:167]: Please add a test reading from a dialect like mysql that has the same NULL propagation semantics as sqlite to verify that the rotation rewrite isn't applied.

**Comment by georgesittas** [comment]
@Kunal8954 any plans to finish this?

**Comment by Kunal8954** [comment]
@georgesittas
Thanks for the feedback. I’ve addressed the suggested changes and added the MySQL regression coverage. All relevant dialect tests, linting, and mypy checks are passing.

File: sqlglot/generators/sqlite.py [hunk 5]
```diff
@@ -338,17 +338,36 @@ def groupconcat_sql(self, expression: exp.GroupConcat) -> str:
...
+        if not expression.args.get("ignore_nulls"):
+            return rename_func(name)(self, expression)
 
-        return self.sql(expression, "this")
+        # GREATEST/LEAST ignore NULLs, but SQLite's multi-argument MAX/MIN
+        # return NULL if any argument is NULL. Emit a rotation of COALESCE
+        # calls so that every argument is the first argument of exactly one
+        # COALESCE, which then returns it if it's non-NULL. Comparing all of
+        # those yields the ignored-NULLs result, and an all-NULL input still
+        # yields NULL:
```
Comment (georgesittas) [inline sqlglot/generators/sqlite.py]: This suffices.

```suggestion
        # SQLite's multi-argument MAX/MIN return NULL if any argument is NULL.
```

**Review by georgesittas — APPROVED** [review]
Thanks for the PR

**Comment by Kunal8954** [comment]
Thanks for reviewing!
